apsuite/optics_analysis/optics_correction.py

```diff
@@ -0,0 +1,217 @@
+        self.cv_idx = self._get_idx(self.fam_data['CV']['index'])
+        self.method = method
+
+    @property
+    def method(self):
+        """."""
+        return self._method
+
+    @method.setter
+    def method(self, value):
+        if value is None:
+            return
+        if isinstance(value, str):
+            self._method = int(value in OpticsCorr.METHODS._fields[1])
+        elif int(value) in OpticsCorr.METHODS:
+            self._method = int(value)
+
+    @property
+    def method_str(self):
+        """."""
+        return OpticsCorr.METHODS._fields[self._method]
+
+    @staticmethod
+    def _get_idx(indcs):
+        return np.array([idx[0] for idx in indcs])
+
+    def calc_jacobian_matrix(self, model=None):
+        """Optics correction response matrix.
+
+        Calculates the variation of Twiss functions and tune given the
+        variation of integrated quadrupoles strength.
+        """
+        if model is None:
+            model = self.model
+
+        res0 = self._get_optics_vector(model)
+        self.jacobian_matrix = np.zeros((res0.size, len(self.knobs_idx)))
+        delta = 1e-6
+
+        modcopy = model[:]
+        for idx, nmag in enumerate(self.knobs_idx):
+            dlt = delta/len(nmag)
+            for seg in nmag:
+                modcopy[seg].KL += dlt
+            res = self._get_optics_vector(modcopy)
+            self.jacobian_matrix[:, idx] = (res-res0)/delta
+            for seg in nmag:
+                modcopy[seg].KL -= dlt
+        return self.jacobian_matrix
+
+    def _get_optics_vector(self, model):
+        twi, *_ = pyaccel.optics.calc_twiss(model)
+        w_beta = 1e2  # order of centimeters
+        w_disp = 1e4  # order of tenth of milimeters
+        w_tune = 1e4  # order of 10^-2
+
+        betax_bpm = twi.betax[self.bpm_idx]
+        betax_ch = twi.betax[self.ch_idx]
+        betay_bpm = twi.betay[self.bpm_idx]
+        betay_cv = twi.betay[self.cv_idx]
+        etax_bpm = twi.etax[self.bpm_idx]
+        tunex = twi.mux[-1]/2/np.pi
+        tuney = twi.muy[-1]/2/np.pi
+
+        vec_bx = np.hstack((betax_bpm, betax_ch))
+        vec_by = np.hstack((betay_bpm, betay_cv))
+        vec_dx = etax_bpm
+        vec_tune = np.hstack((tunex, tuney))
+
+        vec = vec_bx * w_beta / vec_bx.size
+        vec = np.hstack((vec, vec_by * w_beta / vec_by.size))
+        vec = np.hstack((vec, vec_dx * w_disp / vec_dx.size))
+        vec = np.hstack((vec, vec_tune * w_tune / vec_tune.size))
+        return vec
+
+    def get_kl(self, model=None, knobsidx=None):
+        """Return integrated quadrupoles strengths."""
+        if model is None:
+            model = self.model
+        if knobsidx is None:
+            knobsidx = self.knobs_idx
+        kl = []
+        for mag in knobsidx:
+            kl_seg = []
+            for seg in mag:
+                kl_seg.append(model[seg].KL)
+            kl.append(kl_seg)
+        return np.array(kl)
```

Comment:
If you don't sum the KL as commented above. Here you are assuming all knobs have the same number of segments. I know that for the current version of Sirius lattice this is true, but should we make this assumption here?

Reply:
fixed!